### ISSUE TITLE
gradient accumulation

### ISSUE THREAD

**Issue by SoseloX** [volcengine/verl#526]
I'd like to ask whether gradient accumulation can be supported when using GRPO currently. It seems that the actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu needs to be smaller than the train batch size to be effective. Or is there something wrong with my understanding?

![Image](https://github.com/user-attachments/assets/00dd46c1-ee08-48b2-8699-bcda11f2773c)

**Comment by dirtyDan0** [comment]
Advantages(1009) are computed before a batch is splited (1026). They have nothing to do with each other.

https://github.com/volcengine/verl/blob/830fab2aedb7e4cd4fac18e7d1e4830620ee2f1d/verl/trainer/ppo/ray_trainer.py#L986-L1028

**Comment by BeastyZ** [comment]
@dirtyDan0 So may I know how to use gradient accumulation in Verl now?

**Comment by dirtyDan0** [comment]
> [@dirtyDan0](https://github.com/dirtyDan0) So may I know how to use gradient accumulation in Verl now?

Take fsdp backend as example, you can refer to  `update_policy` in `verl/workers/actor/dp_actor.py`. The parameters are updated every mini_batch, inside which you forward a micro_batch for multiple times. You can set the mini_batch to be a multiple of the micro_batch.

**Comment by BeastyZ** [comment]
> > [@dirtyDan0](https://github.com/dirtyDan0) So may I know how to use gradient accumulation in Verl now?
> 
> Take fsdp backend as example, you can refer to `update_policy` in `verl/workers/actor/dp_actor.py`. The parameters are updated every mini_batch, inside which you forward micro_batches for multiple times. You can set the mini_batch to be a multiple of the micro_batch.

Very concise and detailed reply! Thanks a lot.

**Comment by ZetangForward** [comment]
> > [@dirtyDan0](https://github.com/dirtyDan0) So may I know how to use gradient accumulation in Verl now?
> 
> Take fsdp backend as example, you can refer to `update_policy` in `verl/workers/actor/dp_actor.py`. The parameters are updated every mini_batch, inside which you forward micro_batches for multiple times. You can set the mini_batch to be a multiple of the micro_batch.

What about megatron_actor? I notice the megatron_actor utilizes `from megatron.core.pipeline_parallel import get_forward_backward_func` to conduct one fwd_bwd, how to accumulate this gradient?

**Comment by dirtyDan0** [comment]
self.actor_optimizer.step()  is called after self.forward_backward_batch
https://github.com/volcengine/verl/blob/07897f84e5a4f69ca8944ee478c6e6def8962db3/verl/workers/actor/megatron_actor.py#L489-L511
split (normalized mini batch) into micro_batches
https://github.com/volcengine/verl/blob/07897f84e5a4f69ca8944ee478c6e6def8962db3/verl/workers/actor/megatron_actor.py#L281-L311

https://github.com/volcengine/verl/blob/07897f84e5a4f69ca8944ee478c6e6def8962db3/verl/workers/actor/megatron_actor.py#L444-L472

**Comment by ZetangForward** [comment]
> self.actor_optimizer.step() is called after self.forward_backward_batch
> 
> [verl/verl/workers/actor/megatron_actor.py](https://github.com/volcengine/verl/blob/07897f84e5a4f69ca8944ee478c6e6def8962db3/verl/workers/actor/megatron_actor.py#L489-L511)
> 
> Lines 489 to 511 in [07897f8](/volcengine/verl/commit/07897f84e5a4f69ca8944ee478c6e6def8962db3)
> 
>  for data in dataloader: 
>      data.to(torch.cuda.current_device()) 
>      self.actor_optimizer.zero_grad() 
>      # use use_contiguous_buffers_in_local_ddp and no overlap_dp_param_comm 
>      for chunk in self.actor_module: 
>          # if use distributed optimizer, zero grad buffer will be handled by optimizer 
>          chunk.zero_grad_buffer() 
>   
>      calculate_entropy = self.config.entropy_coeff != 0 
>      if data.meta_info.get("micro_batch_size", None) is not None: 
>          micro_batch_size = data.meta_info["micro_batch_size"] 
>      else: 
>          micro_batch_size = self.config.ppo_micro_batch_size_per_gpu 
>      max_token_len = None 
>      if self.config.use_dynamic_bsz: 
>          max_token_len = self.config.ppo_max_token_len_per_gpu * self.config.megatron.context_parallel_size 
>      metric_micro_batch = self.forward_backward_batch(data, calculate_entropy=calculate_entropy, use_dynamic_bsz=self.config.use_dynamic_bsz, micro_batch_size=micro_batch_size, max_token_len=max_token_len, mini_batch_size=self.config.ppo_mini_batch_size) 
>      metric_micro_batch = metric_micro_batch["output"] 
>      for metric in metric_micro_batch: 
>          # Note that o[0] is metrics, o[1] is entropy, o[2] is response_mask 
>          append_to_dict(metrics, metric[0])  # append the metric from this micro-batch to global metrics. 
>   
>      update_successful, grad_norm, num_zeros_in_grad = self.actor_optimizer.step() 
> 
> split (normalized mini batch) into micro_batches
> [verl/verl/workers/actor/megatron_actor.py](https://github.com/volcengine/verl/blob/07897f84e5a4f69ca8944ee478c6e6def8962db3/verl/workers/actor/megatron_actor.py#L281-L311)
> 
> Lines 281 to 311 in [07897f8](/volcengine/verl/commit/07897f84e5a4f69ca8944ee478c6e6def8962db3)
> 
>      def forward_backward_batch(self, data: DataProto, forward_only=False, post_process_fn=None, calculate_entropy=False, use_dynamic_bsz=False, micro_batch_size=None, max_token_len=None, mini_batch_size=None): 
>          """ 
>          We assume: 
>          - The model takes input: (input_ids, attention_mask, position_ids). No rmpad for the input 
>          - The communication shape is (total_nnz_pad_to_sp // tp_size, 1, hidden_size) if sequence parallel is enabled 
>          """ 
>          # broadcast from last pp rank to all other pp ranks 
>          # TODO: actually, we just need to control the sampling order. 
>          mini_batch = data 
>          broadcast_dict_tensor(mini_batch.batch, src=mpu.get_pipeline_model_parallel_last_rank(), group=mpu.get_pipeline_model_parallel_group()) 
>          # split into micro-batches 
>          mini_batch.batch["attention_mask"] = mini_batch.batch["attention_mask"].to(bool) 
>   
>          indices = None 
>          if use_dynamic_bsz: 
>              assert max_token_len is not None, "max_token_len must be set when use_dynamic_bsz is True" 
>              vpp_size = mpu.get_virtual_pipeline_model_parallel_world_size() 
>              if vpp_size is not None and vpp_size > 1: 
>                  microbatch_group_size_per_vp_stage = self.tf_config.microbatch_group_size_per_vp_stage 
>                  micro_batches, indices = rearrange_micro_batches(batch=mini_batch.batch, num_batches_divided_by=microbatch_group_size_per_vp_stage, max_token_len=max_token_len) 
>                  assert len(micro_batches) % self.tf_config.microbatch_group_size_per_vp_stage == 0, f"micro_batches {micro_batches} must be divisible by microbatch_group_size_per_vp_stage {microbatch_group_size_per_vp_stage} for megatron backend" 
>              else: 
>                  micro_batches, indices = rearrange_micro_batches(batch=mini_batch.batch, max_token_len=max_token_len) 
>              total_seqlen = max_token_len 
>          else: 
>              assert micro_batch_size is not None, "micro_batch_size is needed to be passed in when not using dynamic batch size" 
>              micro_batches = mini_batch.batch.split(micro_batch_size) 
>              seq_len = micro_batches[0]["input_ids"].shape[1] 
>              total_seqlen = micro_batch_size * seq_len 
>          # compute input shapes for pp stages 
>          n_micro_batch = len(micro_batches) 
> [verl/verl/workers/actor/megatron_actor.py](https://github.com/volcengine/verl/blob/07897f84e5a4f69ca8944ee478c6e6def8962db3/verl/workers/actor/megatron_actor.py#L444-L472)
> 
> Lines 444 to 472 in [07897f8](/volcengine/verl/commit/07897f84e5a4f69ca8944ee478c6e6def8962db3)
> 
>  batch_generator = make_batch_generator(micro_batches, vpp_size=len(self.actor_module)) 
>   
>  # TODO: we may use the new schedule instead 
>  # for flash-attn: (seq_len, batch_size, hidden_size) = (mbs*seq_len, 1, hidden_size) 
>  if mpu.get_pipeline_model_parallel_world_size() > 1: 
>      losses_reduced = forward_backward_func( 
>          forward_step_func=forward_step, 
>          data_iterator=batch_generator, 
>          model=self.actor_module, 
>          num_microbatches=n_micro_batch, 
>          seq_length=total_seqlen,  # no use when input_shapes was set 
>          micro_batch_size=1,  # no use when input_shapes was set 
>          forward_only=forward_only, 
>      ) 
>  else: 
>      losses_reduced = forward_backward_func( 
>          forward_step_func=forward_step, 
>          data_iterator=batch_generator, 
>          model=self.actor_module, 
>          num_microbatches=n_micro_batch, 
>          seq_length=total_seqlen,  # in use for pp = 1 
>          micro_batch_size=1,  # in use for pp = 1 
>          forward_only=forward_only, 
>      ) 
>  # loss_reduces contains the stats returned from loss_func 
>  losses_reduced = {"output": losses_reduced} 
>  if use_dynamic_bsz: 
>      losses_reduced["indices"] = indices 
>  return losses_reduced

Thx for your reply, I have one more question.

When utilizing gradient accumulation in Megatron (as per the code section you referenced), is it necessary to manually scale the loss by dividing it by the number of accumulation steps at each inner step? 
```python
loss = loss / accumulate_steps
```
Or does Megatron handle gradient averaging automatically by performing this division internally?

@dirtyDan0

**Comment by dirtyDan0** [comment]
https://github.com/NVIDIA/Megatron-LM/blob/6a6cd478839d90cf09a837adf8c79cbc844bc920/megatron/core/pipeline_parallel/schedules.py#L301-L315

**Comment by Kirrito-k423** [comment]
English:

This issue appears answered based on the existing maintainer comments and the current code structure. For FSDP, gradient accumulation is controlled by making `actor_rollout_ref.actor.ppo_mini_batch_size` larger than `actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu`; each mini-batch is split into micro-batches, and the optimizer step happens once after forward/backward over those micro-batches. For Megatron, the engine likewise prepares micro-batches, calls Megatron's `forward_backward_func` with `num_microbatches`, and then performs `optimizer.step()` after the accumulated backward pass. The later Megatron loss-scaling question was also answered by pointing to Megatron-LM's schedule code.

中文：

根据已有维护者回复和当前代码结构，这个问题已经有答案。FSDP 场景下，通过让 `actor_rollout_ref.actor.ppo_mini_batch_size` 大于 `actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu` 来形成梯度累积；每个 mini-batch 会被拆成 micro-batches，并在这些 micro-batches 完成 forward/backward 后再执行一次 optimizer step。Megatron 场景下，engine 同样会准备 micro-batches，使用 Megatron 的 `forward_backward_func` 和 `num_microbatches` 执行前后向，然后在累积后执行 `optimizer.step()`。后续关于 Megatron 是否需要手动缩放 loss 的问题也已由维护者指向 Megatron-LM schedule 代码回答。

This response was generated by a scheduled Codex GPT5.5 issue processing task.
